server/src/main/java/org/apache/calcite/avatica/server/HttpServer.java

```diff
@@ -239,18 +239,8 @@ protected void internalStart() {
     server.setConnectors(new Connector[] { connector });
 
     // Default to using the handler that was passed in
-    final HandlerList handlerList = new HandlerList();
-    Handler avaticaHandler = handler;
-
-    // Wrap the provided handler for security if we made one
-    if (null != securityHandler) {
-      securityHandler.setHandler(handler);
-      avaticaHandler = securityHandler;
-    }
+    configureHandlers(securityHandler);
```

Comment:
At this point, `securityHandler` is null because you don't have a case statement for `CUSTOM` a few lines above.

This means that your CUSTOM AuthenticationType is obtusely "unimplemented". There isn't any clarity in here for me to know what security methods on the `AvaticaServerConfiguration` will actually have effect, and which will not.

For example, I may try to configure the login realm via the `ASC`, but I don't think that would actually be interrogated. When constructing the Jetty server.

Reply:
Agreed, we can add a clause that `CUSTOM` config has to be used with appropriate `ServerCustomizer` to create the server. I can document that appropriately. We can also restrict users to use `ServerCustomizer`  with `CUSTOM` type for the following reason.

`ServerCustomizer` was added to provide additional options for configuration to the server. Since we pass in the `Jetty server` object directly, one can directly remove all the connectors , security handlers and other relevant properties from the server, which can potentially lead to unexpected bugs.

Here (at SFDC), we need a way to customize the servers so that we can configure our own `SSLContextFactory` and `ServerConnector` on the handler (based off some internal code for MTLS auth). Having customizers allow us to do that, however we don't have a mechanism to bypass all default `AvaticaServerConfiguration` logic with the methods provided. We also need to support impersonation, which is only allowed with `buildSpnegoConfiguration()` method, however we don't want to use SPNEGO. This `CUSTOM` method of authentication opens up a lot of doors to customize both authZ and authN as per requirements. As shown in example, we can also use this up for adding impersonation with BasicAuth, which is currently not supported. We still want to use the same handlers as the core logic remains same.

Let me know your thoughts on this one!

I agree that this is not the _best_ way of doing this. I believe that `AvaticaServerConfiguration` interface should provide methods to authenticate the user. For example, the method `isUserPermitted()` is called first inside the `AvaticaProtobufHandler` as well as `AvaticaJsonHandler` classes, where the authentication happens only for the case of SPNEGO. Ideally this method should be wrapped around every handler that is implemented, since I can potentially add a new Handler and forget to call this method at the start. Also, if we add new auth methods, we need to add if-else statements over here to ensure that we cover all the cases. I feel that in such a case `AvaticaServerConfiguration` should be self-sufficient to provide an interface method that can handle the part based on its type. We can use that wrapper to also handle common pieces of code in these handlers like exception handling or impersonation allowance based on serverConfig.

If you can point me to the code that helps in such wrapping, I am happy to do the refactoring.

Reply:
> one can directly remove all the connectors , security handlers and other relevant properties from the server, which can potentially lead to unexpected bugs.

Yeah, this is the angle I'm coming from. I can appreciate the flexibility of `CUSTOM`, but am worried if that's "safe" for the common-folk.

> I believe that AvaticaServerConfiguration interface should provide methods to authenticate the user.

I think that makes sense. My biggest worry about `AvaticaServerConfiguration` is that it's started to become a "dumping ground". Every time I've added a new authn option, it results in an implementation specific option that is only valid "sometimes". Not saying that what you're doing here requires an immediate re-write, just trying to give you my perspective :)

> Ideally this method should be wrapped around every handler that is implemented, since I can potentially add a new Handler and forget to call this method at the start

+1

> If you can point me to the code that helps in such wrapping, I am happy to do the refactoring.

`o.a.c.a.server.HandlerFactory` is what instantiates the `Avatica*Handler` classes now. I think you'd just need to pass the `ASConfiguration` (or some interface which just has the authz/impersonation related methods -- the Avatica handlers shouldn't need to know all HTTP server configuration details) into the construction of the Handler.

This would be some great follow-on.

Reply:
Sure will file a follow up Jira for that.